### ISSUE TITLE
.findOne(undefined) returns first item in the database instead of undefined

### ISSUE THREAD

**Issue by ct-reposit** [typeorm/typeorm#2500]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

```
user.id = undefined;
User.findOne(user.id); // expect to see undefined / null but returns the first item in the table
```

**Comment by pleerock** [comment]
Looks like an issue. Contributions are welcomed!

**Comment by kvnvo** [comment]
Same here, why there is no fix?

**Comment by vlapo** [comment]
Sure from this point it seem like a bug but we have to consider more than one using of `findOne` e.g.:
 - Active record
```
const user = await User.findOne()
```
 - repository
```
const repository = connection.getRepository(User);
const users = await repository.findOne();
```
 - entity manager
```
const user = getManager().findOne(User);
```
All these examples demonstrate using `findOne` with `undefined` value as id input and it is expected return one entity.
Also documentation from sources say:
```
/**
 * Finds first entity that matches given conditions.
 */
findOne(id?: string|number|Date|ObjectID, options?: FindOneOptions<Entity>): Promise<Entity|undefined>;
```
Given condition is nothing (`undefined`) so from my view it is "find one record and I do not have any condition". I think it is all right to return first record in table.

And of course about:
```
User.findOne({ where: { id: null } })
```
It will return undefined because there is no record with `id` = `null`

**Comment by vlapo** [comment]
Is this real issue? I mean `User.findOne(undefined);` is same as `User.findOne();` but what is correct behaviour of `User.findOne();`? 

From jsdoc:
```
/**
 * Finds first entity that matches given conditions.
 */
```
But there are no conditions (`undefined`) and I think if there is no condition we have to return first entity like from `User.findOne({});`

Maybe change method signature and mark id or condition as required but it is really big change and it will be necessary sync at least all `find*` methods behaviour.

**Comment by vlapo** [comment]
But there is function overloading:
```typescript
findOne(id?: string|number|Date|ObjectID, options?: FindOneOptions<Entity>): Promise<Entity|undefined>;
findOne(options?: FindOneOptions<Entity>): Promise<Entity|undefined>;
findOne(conditions?: FindConditions<Entity>, options?: FindOneOptions<Entity>): Promise<Entity|undefined>;
findOne(optionsOrConditions?: string|number|Date|ObjectID|FindOneOptions<Entity>|FindConditions<Entity>, maybeOptions?: FindOneOptions<Entity>): Promise<Entity|undefined> {
        return this.manager.findOne(this.metadata.target, optionsOrConditions as any, maybeOptions);
}
```
You say `undefined` is for first overloaded function and represents id. I say it is possible that `undefined` is for  any other overloaded method.

Your example with `WHERE id = null` can be right in case method name is `findOneById()` but I think `findOne()` is more about `LIMIT 1` part. 
But maybe I am wrong :-) lets wait to @pleerock or another contributors opinion.

**Comment by rustamwin** [comment]
https://github.com/typeorm/typeorm/blob/efb6353fdcdb0715596cc53db48fcf6633d8de7b/src/entity-manager/EntityManager.ts#L580-L621

Probably need check `entityClass` like this
```js
async findOne<Entity>(entityClass: ObjectType<Entity>|EntitySchema<Entity>|string|undefined, idOrOptionsOrConditions?: string|string[]|number|number[]|Date|Date[]|ObjectID|ObjectID[]|FindOneOptions<Entity>|FindConditions<Entity>, maybeOptions?: FindOneOptions<Entity>): Promise<Entity|undefined> { 
  if (!entityClass) {
    return null;
  }
....
```

**Comment by vlapo** [comment]
@rustamwin Did you mean `idOrOptionsOrConditions` instead of `entityClass`?

```typescript
async findOne<Entity>(entityClass: ObjectType<Entity>|EntitySchema<Entity>|string|undefined, idOrOptionsOrConditions?: string|string[]|number|number[]|Date|Date[]|ObjectID|ObjectID[]|FindOneOptions<Entity>|FindConditions<Entity>, maybeOptions?: FindOneOptions<Entity>): Promise<Entity|undefined> { 
  if (! idOrOptionsOrConditions) {
    return undefined;
  }
...
```
This will cause that `User.findOne()` will also return `undefined`. I think this is incorrect behaviour of `findOne`. I would suppose `User.findOne()` will return one record without any conditions (if exists).

**Comment by rustamwin** [comment]
@vlapo i fix this, thnx :smile:

**Comment by rustamwin** [comment]
Perhaps `FindOptionsUtils.isFindOneOptions` need to be improved @pleerock @vlapo

**Comment by vlapo** [comment]
From my point there is nothing to improve and `findOne` works as intended.

**Comment by pleerock** [comment]
It looks like we either need to left it as it is, either to enforce `findOne` to accept some real conditions. What do you think @Kononnable @havenchyk ?

**Comment by havenchyk** [comment]
@pleerock I personally would add a check if any value was passed or not

**Comment by vlapo** [comment]
@havenchyk but current method signatures allow this case `User.findOne()` and this check will throw error. Of course workaround is `User.findOne({})`. But in that case we should change method signature and require at least one parameter.

**Comment by havenchyk** [comment]
@vlapo why do you think it will throw an error?

just naive code:
`...args => args.length > 0 ? args[0] === undefined && doSomething() : returnFirstItem()`

of course, I see in code base it's much more complicated, but I don't the area for an error. Could you please explain, what I have missed?

**Comment by vlapo** [comment]
Sry I see you want to throw error in previous comment. My mistake :-) 

So you vote for `findOne()` (without args) will return undefined. Not first one?

**Comment by havenchyk** [comment]
Seems I wasn't clear enough. Let me try once again.

If argument is passed and it's value is not undefined, handle this value.
If agrument is passed and it's value is undefined, return undefined/throw an exception
if agrument is not passed, return the first row.

**Comment by vlapo** [comment]
Understand now. I think it should work 👍

**Comment by pleerock** [comment]
> If agrument is passed and it's value is undefined, return undefined/throw an exception

choose one. Return undefined or throw an exception?)

**Comment by Kononnable** [comment]
Sorry for being little late for the party.

>No, you except same behavior as SQL
SELECT * FROM users WHERE id = null LIMIT 1

I completely disagree. Passing undefined shouldn't cause passing null to sql query. Null and undefined are completely different primitive values. Undefined should be ignored(optional parameters) or always return no records(without querying).

I think if we want to change current behavior we should throw an exception. If someone is deliberately passing value of undefined we can safely assume that he doesn't really know what he is doing and there is a human error somewhere. Returning undefined will just populate logical error to later called functions(app logic).

We can't search for something if we don't know what are we looking for - it means that there is a logical error while defining what are we looking for.

**Comment by mazyvan** [comment]
I'm having the same issue with `findOneOrFail` IMHO returning the first element when passing `undefined` as a parameter must not be the desired behavior. It's dangerous and no other ORM that I know does that

**Comment by matiasbeltramone** [comment]
The same issue for findOne and findOneOrFail, it's complete dangerous if you passing undefined as a parameter, must not be return the first element. :(

**Comment by Blockost** [comment]
> Sorry for being little late for the party.
> 
> > No, you except same behavior as SQL
> > SELECT * FROM users WHERE id = null LIMIT 1
> 
> I completely disagree. Passing undefined shouldn't cause passing null to sql query. Null and undefined are completely different primitive values. Undefined should be ignored(optional parameters) or always return no records(without querying).

@Kononnable Well, the same behaviour happens when calling `findOneOrFail(null)`: first row is returned

**Comment by mazyvan** [comment]
I think this could be fixed just adding a special configuration. So any user could decide by its own if he/she wants the findOne method to return the first item when passing an undefined value

**Comment by malimccalla** [comment]
This had me scratching my head for sooo long today! I wrongly assumed that `findOne(userId)` where `userId` is `undefined` would not return me a user. Is this a bug?

**Comment by uPaymeiFixit** [comment]
This caused several hours of debugging for me today as well. Specifically with the method `findOneOrFail` (also affected). The name suggests the method will fail if it can't find a matching id in the database, like when the search target is null. This unexpected behavior seems to be causing a headache for many typeorm users.

I cannot find a scenario where a developer would expect to pass in undefined and receive the first row. While it's possible somebody is intentionally using the method this way, I would guess there are far more developers stumped by this.

**Comment by malimccalla** [comment]
> If agrument is passed and it's value is undefined, return undefined/throw an exception

IMO this should be the default for `findOne` having it return undefined and `findOneOrFail` throw an exception.

**Comment by mazyvan** [comment]
I really don't understand why this is still unresolved I totally agree with @uPaymeiFixit Let's be clear. NOBODY WANTS THE FINDONE ALONE TO RETURN THE FIRST ROW. This is making more harm than good.

I understand you guys don't want to add a breaking change (IMHO it is a fix, but...) so please please consider just to add configuration property to change this default behavior

**Comment by LuisAngelMoralesRomo** [comment]
> I really don't understand why this is still unresolved I totally agree with @uPaymeiFixit Let's be clear. NOBODY WANTS THE FINDONE ALONE TO RETURN THE FIRST ROW. This is making more harm than good.

I do @mazyvan
The [documentation](https://typeorm.io/#/entity-manager-api) is clear:

> findOne - Finds the first entity that matches some id or find options.

Given no conditions, first row matches.

I personally like and use @malimccalla recomendation:

>Worth mentioning though that if you wish to avoid this you could just do .findOne({ where: { id } }) in which case undefined would be returned

It is less ambiguous and more verbose than the function overload.

**Comment by mazyvan** [comment]
The fact that the docs are right doesn't mean this behavior is useful. 

Tell me @LuisAngelMoralesRomo are you really going to use the `findOne` method alone to get the fist record on your database table? If not I can't see the usefulness of this. Even if you say you are going to use it, I'm sure you ended up using `.findOne({ where: { id } })` more.

And even all of this. Don't you really really think that someone else is going to give problems this?

The reason many of us ended up here is because we get some troubles with this right? And if this stays the same more people will be facing this too.

**Comment by LuisAngelMoralesRomo** [comment]
You are right @mazyvan, this issue exist for a reason.

>Tell me @LuisAngelMoralesRomo are you really going to use the findOne method alone to get the fist record on your database table? If not I can't see the usefulness of this. Even if you say you are going to use it, I'm sure you ended up using .findOne({ where: { id } }) more.

You are right, I have used it only a couple of times, and it is not that useful. And yes I use it in the way you wrote.

Maybe I'm a little obsessive with the docs. I just wrote my opinion if they decide to make a change. I personally use ` .findOne()` more like a `find({ take: 1 })` to get the row directly instead of an array, not to query by id.

If you ask me, I prefer less function overloading, I would keep just this one:
```ts
findOne(options?: FindOneOptions<Entity>): Promise<Entity|undefined>;
```
And that is how I use it.

BTW, I'm from Zapopan too.

**Comment by tjhillard** [comment]
> Seems I wasn't clear enough. Let me try once again.
> 
> If argument is passed and it's value is not undefined, handle this value.
> If agrument is passed and it's value is undefined, return undefined/throw an exception
> if agrument is not passed, return the first row.

@havenchyk  How do you determine if an optional argument is passed but the value is `undefined` when it's default value is `undefined`, as per the norm with optional arguments?

I started working on a PR for this fix and have the behavior how I think it should be, but after reading through some of the tests around `findOne` and `findOneOrFail`, the fact they return the first record is indeed intentional behavior, thus this not really being a "fix", but unfortunately a "change", and a breaking one. 

>test/repository/find-methods/repository-find-methods.ts:356
```ts
it("should return first when no criteria given", () => ...
```

IMO, I agree with others in the thread that the current behavior for this method is dangerous and misleading, especially with `findOneOrFail(undefined)` returning a record.

What if, similarly to ActiveRecord and others, we introduce two new methods:

```ts
repository.first(); // Returns first record
repository.last(); // Returns last record
```

This allows users to have a simple and explicit way of receiving the first and last records and makes the find methods do only what they should do... find things based on given criteria, and not implicitly fallback to returning the first record.

The rollout could be:

1. Introduce new additive first() and last() methods
2. Add deprecation warnings to findOne and findOneOrFail when they are called with undefined find criteria, and encourage the use of `repository.first()`
3. Eventually, implement the breaking change (fix) for returning undefined/throwing exception when given undefined or no arguments.

**Comment by Mcfloy** [comment]
> > Seems I wasn't clear enough. Let me try once again.
> > If argument is passed and it's value is not undefined, handle this value.
> > If agrument is passed and it's value is undefined, return undefined/throw an exception
> > if agrument is not passed, return the first row.
> 
> @havenchyk How do you determine if an optional argument is passed but the value is `undefined` when it's default value is `undefined`, as per the norm with optional arguments?

I think he meant `if argument is passed and it's value is null, return the first row` as null has its own primitive type, you can use it do determine that the user asked specifically that he had no condition.

Problem on the deprecation warning is that it wouldn't prevent someone to assign a value to a either `any `or an `object` type, so you could have undefined value on runtime and still not getting the warning, or you would throw an exception.

**Comment by pleerock** [comment]
okay, so for me `findOne` without any criteria never make sense. So I'm okay with deprecating this signature. I don't see point in `first` and `last` methods either - databases sometimes give data in not expected order and this may bring confusion. 

so, lets deprecate `findOne` without arguments and throw error if first argument passed is null or undefined. Feel free to PR.

**Comment by mlandisbqs** [comment]
Is there a case where undefined options is actually valid to return a single record?  the age old, ``select 1 from dual`` comes to mind. also there could be valid use cases where a table stores one and only record - no where clause needed. I came here because I had a bad bug in my code where I forgot to pass the id as criteria for a user query and the findOne code was silently returning the system user every time (dev only) :( In that case I was still passing options to traverse a relation so the deprecated fix would not have saved me - methinks.

Not sure how to fix this without a breaking change - Maybe deprecate findOne and spit out a big warning?  Provide an alternate function "findExactlyOne(OrFail)" that checks the response for multiple records?

**Comment by gempain** [comment]
This IMO is a security issue that could lead users to leak data by mistake. Particularly if you're using GraphQL resolvers which have gone past the HTTP layer validation.

**Comment by pleerock** [comment]
Since this is a breaking, fix will be available in the `next` branch.

**Comment by elovin** [comment]
the same problem exists with the delete method, which is even worse IMO

**Comment by AbuOmar** [comment]
I'll give an example where this could be very dangerous:

- Imagine you have a JWT payload that includes the user id like {userId: 5} with a signature 
- The developers in back-end for some reason changed "userId" to "id" in the payload like {id: 5}
- Now the server should see old JWT payloads like {userId: 5} as invalid, but instead of that when using User.findOne(payload.id} (payload.id in undefined in old payloads), and because of that: old payloads will give access as the first user in database who is usually a super admin.

**Comment by jzfgo** [comment]
This is still broken as of `v0.2.37`.

**Comment by jmls** [comment]
man, I've just spent several hours trying to figure out what the hell was going on with my pos code. This. This is the problem.

(a null id was being passed to findOne(id) )

ouch. I vote for removing this dangerous "feature"

When is the "next" branch available ?

**Comment by bhalvy** [comment]
This has just caused a very serious issue in my application when using `.findOneOrFail()`. I cannot see any situation where an undefined or null in this method should result in anything but an error.

**Comment by jorgestg** [comment]
Shouldn't passing `undefined` ignore the field rather than passing `null` to the SQL?

```js
User.find({ someField: undefined });
```

should translate into 

```sql 
SELECT * FROM user;
```

but now it translates to

```sql
SELECT * FROM user where someField = null;
```

isn't that the whole point of having `undefined` and `null`?

**Comment by brianadams731** [comment]
I just spent an hour pulling through my code trying to figure out what was wrong. I honestly thought that the idea of returning items from the db without explicitly passing their ID seemed so ridiculous I didn't bother to check until it was my last option. Returning the first item in the table when undefined is passed to find or findOne is not something I would have (or anyone would have) intuitively expected nor is it documented anywhere. This should not be considered a feature, and should be documented in big bold letters in the find option documentation. This behavior can cause hard to track bugs, and can potentially leak data.

**Comment by darrenbarklie** [comment]
I also agree that this is unintuitive behaviour and should be reconsidered. Why would "defining nothing" be anticipated to "return something"?  Should be documented at a minimum.

**Comment by nhhockeyplayer** [comment]
can someone kind enough post the one line code solution

**Comment by satc0ride** [comment]
> Sorry for being little late for the party.
> 
> > No, you except same behavior as SQL
> > SELECT * FROM users WHERE id = null LIMIT 1
> 
> I completely disagree. Passing undefined shouldn't cause passing null to sql query. Null and undefined are completely different primitive values. Undefined should be ignored(optional parameters) or always return no records(without querying).
> 
> I think if we want to change current behavior we should throw an exception. If someone is deliberately passing value of undefined we can safely assume that he doesn't really know what he is doing and there is a human error somewhere. Returning undefined will just populate logical error to later called functions(app logic).
> 
> We can't search for something if we don't know what are we looking for - it means that there is a logical error while defining search criteria.

I agree with this, semantically undefined is a property that not exist, and null is a property with not value.

I solved it with { id ?? null }

**Comment by lloiacono** [comment]
> 

It's still broken on `0.3.6`, what is the current status with this issue?

**Comment by bastianschott** [comment]
Just to raise more awareness here, still broken in version `0.3.7`

This can lead to major security issues, like @AbuOmar mentioned.

**Comment by crazyoptimist** [comment]
`findOne` is deprecated and `0.4.0` have come up with `findOneBy`. 😢

**Comment by jacobswain** [comment]
Still broken.  This is a fundamental issue with the framework that breaks conventions, introduces massive security issues, requires defensive programming practices, and ultimately calls into question the integrity of the framework.

**Comment by crazyoptimist** [comment]
Yes seriously 😞 
Why was this issue even closed??

**Comment by crisvilla-oss** [comment]
I had the same issue, till I tried adding the @Param decorator to the params given to the function call. 

user.controller
```
@Get('users/:id')
  async findId(@Param('id') id: number): Promise<User> {
    return await this.userService.findById(id);
  }
```
  
user.service
```
async findById(idB: number): Promise<User> {
    const user = await this.usersRepository.findOneBy({ id: idB });
    if (!user) {
      const errors = { User: ' not found' };
      throw new HttpException({ errors }, 404);
    }
    return user;
  }
```

Having done that, it works correctly!

**Comment by dill0wn** [comment]
This is bonkers. I think I am running into the same issue. I am currently running this query
```
const m = await Model.findOneBy({ id: null })
```
And the result is the first row in the table  (postgresql) which happens to have `{ id: 1 }`. This is entirely unexpected behavior.

I would expect an empty result (null), or an exception.

**Comment by juliarnasution** [comment]
Why is this issue closed?

**Comment by Vimbi** [comment]
I consider this behavior unacceptable. It is necessary to make changes, or at least make a warning about such behavior in the documentation. Closing the question does not fix the problem.

**Comment by Johalternate** [comment]
As a workaround you do this:
```
User.findOne({ where: { propertyName: Equal(null) } }) //returns null
User.findOne({ where: { propertyName: Equal(undefined) } }) //returns null
User.findOne({ where: { propertyName: Equal('not found') } }) //returns null
```

**Comment by Sid-Turner-Ellis** [comment]
Even:

```
findOne({
  where: {
    id: null
  }
})

```
Is this definitely expected behaviour?

**Comment by drowzeenico** [comment]
This issue marked as "closed", so I was glad at first. Until I found out the problem is still here. More than that, neither 
`findOneBy({ id: null })`  nor `findOne({
    where: {
        id: null
    }
})`  work with nullable id and return first row in the table.

But what happened? I used to work with TypeORM few years ago and this methods worked fine.

**Comment by Sid-Turner-Ellis** [comment]
> > Even:
> > ```
> > findOne({
> >   where: {
> >     id: null
> >   }
> > })
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > returns the first item for me, is this definitely expected behavior?
> 
> I believe it is, it would make sense if you think about using a function with multiple parameters. Say you have a filter in the front end which is a form with 5 parameters, you can just use the filter dto as an argument for the function without having to worry about empty values. If the behavior were different, then you would need to handle every null/undefined case within the function. Instead, you can just use `Equal(value)` wherever you won't allow nulls to be ignored.

Hmm I see, is this default case though? 

This basically renders all find methods on repositories completely useless for me and I don't think my case is anything but standard. 


Using Equal is cool but it doesn't allow you to pass undefined values to it from what I can tell as it gives you a type error so doesn't really help in my case. At the moment i'm just going to do the check myself which I don't like at all 

```
export const failFindId = <T>(value: T): string =>
  typeof value === "string" ? value : "-1";

const entity = await dataSource.getRespository(User).findOne({where: { id: failFindId(idOrUndefined) }})

```
For my application I know that it should work but I don't like it at all.

**Comment by ssilve1989** [comment]
This is a major problem preventing upgrading to the `0.3.x` releases. What is going on here?

**Comment by Sid-Turner-Ellis** [comment]
Issue created: https://github.com/typeorm/typeorm/issues/9446

**Comment by erencay** [comment]
I was happy with other ORMs but I had to go back for a legacy project and faced with this issue, again, after years. When I saw 0.3 release, I was hopping for them to change this behavior. Indeed they did, but -unsurprisingly- made it worse.

**Comment by LuckyArd110** [comment]
whats happen now? hmm

**Comment by Sid-Turner-Ellis** [comment]
@LuckyArd110 I've created an issue which was apparently a dupe - https://github.com/typeorm/typeorm/issues/9316

**Comment by DanielMaranhao** [comment]
I want to share something that happened with me in my workplace. Thankfully, it was not something very serious, but due to this bug(?), it could be far worse.

We made a change in the JWT payload (changed `userId` field to `sub`). I didn't predict this behavior, but probably because the secret didn't change, the current tokens of the people using the system were still valid. So they still entered in a check that, even though the token is valid, if the user still exists. So the `sub` field was used to fetch the user.

So exactly this happened: as the old tokens still had `userId` and not `sub`, it was a `findOne(undefined)`. The first record was retrieved. But, the first record was Admin. In the end, everyone using the system would become Admin.

Fortunately, we noticed this quickly and fixed it. Also, only internal people we using it, and no one was at this specific moment. But just to show that this bug(?) of TypeORM could have, it seems, devastating effects. Of course, I know that we commited the mistake while altering the JWT contents, but if just nothing was retrieved, instead of the first record, this situation would have been avoided.

**Comment by LaLaLaOpera** [comment]
Even It should have been our responsibility to check and make sure null or undefined parameter never go though query in a first place, but it is disappointing to see that the ORM still can't handle a  simple exception like this.

**Comment by Chris4589** [comment]
con Equal(value) pude solucionar el error.

**Comment by issar13** [comment]
> con Equal(value) pude solucionar el error.

doesn't this break other parts of an application?

**Comment by Chris4589** [comment]
> 

no

**Comment by avilabiel** [comment]
TypeORM should be responsible to return the right data whenever there is a nullable or undefined value in `findOne()` function. This is weird and could be really dangerous depending on your application.

In my case, it returned a random value and we made our data dirty. It took a whole day to fix the messy data caused by this bug.

**Comment by issar13** [comment]
jst use the equal option on typeorm, fixes the issue

**Comment by avilabiel** [comment]
> jst use the equal option on typeorm, fixes the issue

That does not make sense. If I'm finding one by `id: null`, the logical response would be bringing `null`, not the 1st resource/row.

**Comment by zmts** [comment]
v0.3.12 same issue ((

**Comment by TheRaLabs** [comment]
fix this asap

**Comment by DaviRJ** [comment]
The current behavior of this functionality is far from being even what appears in the TypeORM documentation. It needs to be fixed asap.

![image](https://github.com/typeorm/typeorm/assets/28058907/9c2aa512-49b8-4f1a-aa2d-16a7f53375de)

**Comment by Irungaray** [comment]
Spent a couple hours debugging this. The frontend was not sending the id on the body (as it was meant to), so .findOne was returning the first user, that was a super admin as @AbuOmar said.

Thankfully we caught it on a development enviroment, otherwise it could've been a serious security issue.

**Comment by HaiAlison** [comment]
in my case, I will do this:
```
import { IsNull } from 'typeorm';

const barn = await BarnsModel.findOneBy({
      id: dto.barn_id || IsNull()
    });

**Comment by brunomichalski** [comment]
> in my case, I will do this:
> 
> ```
> import { IsNull } from 'typeorm';
> 
> const barn = await BarnsModel.findOneBy({
>       id: dto.barn_id || IsNull()
>     });
> ```


This works for me, but I don't think it should work this way

**Comment by HaiAlison** [comment]
> t

yes, it works when you ensure that the field does not have any null value

**Comment by HaiAlison** [comment]
> > in my case, I will do this:
> > ```
> > import { IsNull } from 'typeorm';
> > 
> > const barn = await BarnsModel.findOneBy({
> >       id: dto.barn_id || IsNull()
> >     });
> > ```
> 
> This works for me, but I don't think it should work this way

yes, it works when you ensure that the field does not have any null value

**Comment by ofuochi** [comment]
> .findOne({ where: { id } })

This has the same behaviour as returning the first record

**Comment by ezescigo** [comment]
So what would be the best practice to use findOne? or should we use other method?

**Comment by amroczeK** [comment]
I am using `"typeorm": "^0.3.17"` and am performing the following Query:

```
const chargeDecisionRequest =
        await this.chargeDecisionRequestRepository.findOne({
          where: { transaction_id: data.transactionId },
        });
```

If I supply a transactionId of "IMI0:2:52:4398082795841" I am being returned the first entry in the database where the transaction id doesn't match. This does not occur when I perform a raw query e.g. `SELECT * FROM charge_decision_request WHERE transaction_id = 'IMI0:2:52:4398082795841'`

My database only has one record:
```
{
  "id": "546963ec-9b51-4436-b137-ce30d024a69a",
  "transaction_id": "IMI0:2:52:4398082795846",
  "imsi": "505013469157558",
  "s_nssai": "200001",
  "charge_period": "3600",
  "charge_start_timestamp": "1684340390",
  "status": "processing",
  "created_at": "2023-10-26T02:35:28.716Z",
  "updated_at": "2023-10-26T02:35:28.716Z"
}
```

**Comment by os-moussao** [comment]
> So what would be the best practice to use findOne? or should we use other method?

Imo, using `.createQueryBuilder()` is the most convenient alternative to `findOne()` so far.

**Comment by mariopepe** [comment]
No way this must be a meme

**Comment by Alamonall** [comment]
LMAO. When is this will be fixed????

**Comment by landersonalmeida** [comment]
> LMAO. When is this will be fixed????

Maybe in 2024 🤣

**Comment by issar13** [comment]
> > LMAO. When is this will be fixed????
> 
> Maybe in 2024 🤣

from 2018?.....no hope

**Comment by issar13** [comment]
everyone just use Equal()

**Comment by merthanmerter** [comment]
wow. thats a dangerous bug! and yes, it is a bug!

**Comment by issar13** [comment]
> wow. thats a dangerous bug! and yes, it is a bug!

but wait there's more...apparently when you query findOne. It makes an sql query twice. turn logging on and see it in realtime.

**Comment by merthanmerter** [comment]
> This is a huge security risk, and not the way it should work. Until PR #9487 is merged, I have created an eslint rule that warns about it.
> 
> ```
> rules: {
>     'no-restricted-syntax': [
>       'error',
>       {
>         selector:
>           "ObjectExpression > .properties[key.name='where']:has(.value.properties > .value:not(CallExpression))",
>         message:
>           'Unsafe where condition, that can leak data. Use Equal() instead.'
>       }
>     ],
> }
> ```

Thank you a lot. To use it with nested object like the example, I modified as below.

`where: { customer: { id: Equal(id) }, status: Equal(Status.ACTIVE) },`

```
'no-restricted-syntax': [
      'error',
      {
        selector:
          "ObjectExpression > .properties[key.name='where'] > .value > .properties:not(:has(CallExpression)), ObjectExpression > .properties[key.name='where'] > .value > .properties > .value > .properties:not(:has(CallExpression))",
        message: 'Unsafe where condition, that can leak data. Use a proper function call instead.',
      },
    ],
```

**Comment by YuriiBoikoOpn** [comment]
just leave here
so issue not just with 1 property in `findOne`
but also with another cases: `findOne`, `findOneBy`,  `findBy` when there are many properties
example:
`const patientId=120`
`const id= null`

`repository.findBy({id, patientId})` - will return not expected all raws `where patientId=120` , but will skip `id` validate at all (like `and id = undefined` not exist )

more tested examples on local:
<img width="1268" alt="Monosnap appointment-availability-milestone-alert test ts — pollin-backend 2023-12-15 21-08-32(1)" src="https://github.com/typeorm/typeorm/assets/93529463/f49e153e-314b-42a9-93c5-734645424e9c">

**Comment by twigs67** [comment]
Can't believe this is still an issue. Would this not be an issue for `update()` and `delete()` as well? @YuriiBoikoOpn would you happen to know this?

**Comment by mistryaksh** [comment]
This is not a proper method to get the data but we can also use this method to findOne in typeorm mongodb

 ```
 const code = await systemCode.find({});
        const data = code.find((props) => {
          if (props._id.toString() === req.params.systemCodeId) {
            return props;
          }
        });
        return Ok(res, data);
```

**Comment by Pyakz** [comment]
this also happens when i pass a  non existing uuid, it returns the first item always

**Comment by JJBocanegra** [comment]
We are about to enter 2025 and this BIG AS A MOUNTAIN BUG is still happening! 🐛 🐞

**Comment by tsvetan-vr** [comment]
> > This is a huge security risk, and not the way it should work. Until PR #9487 is merged, I have created an eslint rule that warns about it.
> > ```
> > rules: {
> >     'no-restricted-syntax': [
> >       'error',
> >       {
> >         selector:
> >           "ObjectExpression > .properties[key.name='where']:has(.value.properties > .value:not(CallExpression))",
> >         message:
> >           'Unsafe where condition, that can leak data. Use Equal() instead.'
> >       }
> >     ],
> > }
> > ```
> 
> Thank you a lot. To use it with nested object like the example, I modified as below.
> 
> `where: { customer: { id: Equal(id) }, status: Equal(Status.ACTIVE) },`
> 
> ```
> 'no-restricted-syntax': [
>       'error',
>       {
>         selector:
>           "ObjectExpression > .properties[key.name='where'] > .value > .properties:not(:has(CallExpression)), ObjectExpression > .properties[key.name='where'] > .value > .properties > .value > .properties:not(:has(CallExpression))",
>         message: 'Unsafe where condition, that can leak data. Use Equal() instead.',
>       },
>     ],
> ```

Thanks. I've extended on this to also include findOneBy method as it also produces unwanted results without using Equal method:

```
    'no-restricted-syntax': [
      'error',
      {
        selector:
          "CallExpression[callee.property.name='findOneBy'] > ObjectExpression > .properties:not(:has(CallExpression))",
        message:
          'Unsafe condition in findOneBy, that can leak data. Use Equal() instead.',
      },
      {
        selector:
          "ObjectExpression > .properties[key.name='where'] > .value > .properties:not(:has(CallExpression)), ObjectExpression > .properties[key.name='where'] > .value > .properties > .value > .properties:not(:has(CallExpression))",
        message:
          'Unsafe where condition, that can leak data. Use Equal() instead.',
      },
    ],
```
![image](https://github.com/user-attachments/assets/ca2451d3-1135-4b35-abff-f137ef8559dc)

**Comment by ezescigo** [comment]
Why is this closed? isn't considered a bug?

**Comment by JJBocanegra** [comment]
Sorry for the double comment, but I continued to modify the rule and added support for allowing numbers and also to restrict the `where: {}` check to not apply to any object that contains the property `where`, but only to the ones that are parameters of one of TypeORM methods.

```
'no-restricted-syntax': [
        'error',
        {
          selector:
            "CallExpression[callee.property.name=/^(findOneBy|findOneByOrFail|findBy|findAndCountBy|)$/] > ObjectExpression > .properties:not(:has(CallExpression)):not([value.value=true]):not([value.value=false]):not([value.type='Literal'][value.raw=/^-?\\d+(\\.\\d+)?$/])",
          message:
            'Unsafe condition that can leak data. Use Equal() instead.',
        },
        {
          selector:
            "CallExpression[callee.property.name=/^(findOne|findOneOrFail|find|findAndCount)$/] > ObjectExpression > .properties[key.name='where'] > .value > .properties:not(:has(CallExpression)):not([value.value=true]):not([value.value=false]):not([value.type='Literal'][value.raw=/^-?\\d+(\\.\\d+)?$/]), CallExpression[callee.property.name=/^(findOne|findOneOrFail|find|findAndCount)$/] > ObjectExpression > .properties[key.name='where'] > .value > .properties > .value > .properties:not(:has(CallExpression)):not([value.value=true]):not([value.value=false]):not([value.type='Literal'][value.raw=/^-?\\d+(\\.\\d+)?$/])",
          message:
            'Unsafe where condition that can leak data. Use Equal() instead.',
        },
      ],
```

**Comment by michaelbromley** [comment]
Fixed in v0.3.27: https://typeorm.io/docs/data-source/null-and-undefined-handling/
